chore: updates compgen templates for new interfaces

jburns24 opened the pull request:
Generated a new receiver found red squigglies, updated the templates. Regenerated and verified the linter was happy. Ran mdatagen on the generated receiver and verified that mdata gen was generating the new interfaces.

adrielp commented:
Looks right @jburns24  - did you run the utility to check everything went well?

acramsay commented:
[From this commit:](https://github.com/open-telemetry/opentelemetry-collector/commit/bd784f03e26a478324175f72fe1e528bfb397d57#diff-48634bb50fbc5a45681e7932facb0a7d896f474caa84c7f41066bc57bc47ea36L50)
> // Deprecated: [v0.103.0] Use receiver.Settings instead.

Nice find and thanks for fixing @jburns24! LGTM but I agree with @adrielp that we should run a generated receiver before merging. A generated receiver should start up successfully and do nothing at all.

jburns24 commented:
😏 I did not actually run it. Ill update the ticket with the results of actually testing the change!

adrielp commented:
This ready to be reviewed again @jburns24 ?

jburns24 commented:
@adrielp This is ready to be reviewed again.

jburns24 commented on Makefile.Common:
```diff
@@ -74,3 +74,6 @@ fmt:
 scan:
 	$(OSV) -r .
 
+.PHONY: mdatagen
+mdatagen:
+	$(MDATAGEN) metadata.yaml
```
I am not sure the history but compgen README referenced a makefile target for mdatagen. It is not there that I could find

acramsay replied:
ok, this one is rather complicated. But the TL;DR is that I think we can revert this change.

Here's how it works:

* Run compgen to build a new component
* You now have a Makefile rendered from this [template](https://github.com/liatrio/liatrio-otel-collector/blob/main/cmd/compgen/cmd/receiver/templates/Makefile.tmpl) which imports the top-level Makefile.Common
* Makefile.Common defines the [gen](https://github.com/liatrio/liatrio-otel-collector/blob/main/Makefile.Common#L62-L63) target as `go generate ./...` which runs at the root dir of the new component
* `go generate ./...` executes mdatagen based on the comment in the generated [doc.go](https://github.com/liatrio/liatrio-otel-collector/blob/main/cmd/compgen/cmd/receiver/templates/doc.go.tmpl#L4) file

@jburns24 can we test this by generating a new dummy receiver and running `make gen` from the root of the new component's repo? As you already mentioned in our call, the relative path to the mdatagen binary is likely wrong. Nice find!!

adrielp replied:
I'd opt for not adding back to the `mdatagen` rule target in the makefile, and fixing the doc instead.

jburns24 replied:
I thought I pushed up these changes 🤦

jburns24 commented on cmd/compgen/cmd/receiver/templates/config_test.go.tmpl:
```diff
@@ -28,23 +29,23 @@ custom config for our custom receiver.
 	type fields struct { // Custom struct for the Config passed into Unmarshal() by reference
 		ControllerConfig scraperhelper.ControllerConfig
-		sample           string
+		Sample           string  `mapstructure:"sample"`
```
The changes of casing `Sample` and mapstructure data resolved a few of the failing tests that were giving this error `decoding failed due to the following error(s): '' has invalid keys: sample`

jburns24 commented on cmd/compgen/cmd/receiver/templates/config.go.tmpl:
```diff
@@ -14,16 +14,16 @@ import (
-	ErrMustNotNil    = errors.New("sample interface must not be nil")
-	ErrSampleConfig  = errors.New("sample config data is required")
-	ErrMustLowercase = errors.New("sample config data must be lowercase")
+	ErrMustString    = "'sample' expected type 'string'"
+	ErrMustNotNil    = "sample interface must not be nil"
+	ErrSampleConfig  = "sample config data is required"
+	ErrMustLowercase = "sample config data must be lowercase"
```
The switch to error strings was because [this test](https://github.com/liatrio/liatrio-otel-collector/blob/7839d171c2e7f4d9fe58a8733f6021f8f8e9fcf7/cmd/compgen/cmd/receiver/templates/config_test.go.tmpl#L69-L76) was returning an error from [here](https://github.com/liatrio/liatrio-otel-collector/blob/7839d171c2e7f4d9fe58a8733f6021f8f8e9fcf7/cmd/compgen/cmd/receiver/templates/config.go.tmpl#L36). confmap.Unmarshall was bubbling the error up from mapstructure.Decode which did not have any public error variables for comparison.

jburns24 commented:
@acramsay @adrielp Also verified the receiver runs

```
2024-08-21T09:44:44.579-0700    info    internal/scraper.go:52  running the myreceiver scrape function  {"kind": "receiver", "name": "myreceiver", "data_type": "metrics"}
2024-08-21T09:44:44.581-0700    info    MetricsExporter {"kind": "exporter", "data_type": "metrics", "name": "debug", "resource metrics": 0, "metrics": 0, "data points": 0}
2024-08-21T09:44:54.580-0700    info    internal/scraper.go:52  running the myreceiver scrape function  {"kind": "receiver", "name": "myreceiver", "data_type": "metrics"}
2024-08-21T09:44:54.580-0700    info    MetricsExporter {"kind": "exporter", "data_type": "metrics", "name": "debug", "resource metrics": 0, "metrics": 0, "data points": 0}
^C2024-08-21T09:45:01.018-0700  info    otelcol@v0.107.0/collector.go:318       Received signal from OS {"signal": "interrupt"}
2024-08-21T09:45:01.018-0700    info    service@v0.107.0/service.go:258 Starting shutdown...
2024-08-21T09:45:01.018-0700    info    healthcheck/handler.go:132      Health Check state change       {"kind": "extension", "name": "health_check", "status": "unavailable"}
2024-08-21T09:45:01.018-0700    info    extensions/extensions.go:63     Stopping extensions...
2024-08-21T09:45:01.018-0700    info    zpagesextension@v0.107.0/zpagesextension.go:105 Unregistered zPages span processor on tracer provider   {"kind": "extension", "name": "zpages"}
2024-08-21T09:45:01.018-0700    info    service@v0.107.0/service.go:272 Shutdown complete.
```

jburns24 commented:
@acramsay Please give this another 👀. I removed mdatagen make target. Fixed the path to mdatagen in the go generate annotation and generated a new receiver and tested that `make gen` worked and the receiver worked.

adrielp commented on cmd/compgen/README.md:
```diff
@@ -15,6 +15,14 @@ types are currently supported:
 
+## Installing `compgen`
+
+From within `cmd/compgen` run `go install .`. This will make build the `compgen`
```
We should have a ticket to fix the name of this so that it doesn't conflict with zsh.

jburns24 replied:
https://github.com/orgs/liatrio/projects/25/views/1?pane=issue&itemId=76028102

adrielp commented on cmd/compgen/README.md:
```diff
@@ -59,7 +67,8 @@ to generate aditional code for the component.
 
-You can run mdatagen by running `make gen`.
+Assuming you have uncommented the include of Makefile.Common in the generated
+receiver, you can run mdatagen by with `make mdatagen`.
```
Run `make generate` instead

acramsay replied:
This is a good callout actually but I think it warrants a little more explaining.

`make gen` is defined in the [Makefile.Common](https://github.com/liatrio/liatrio-otel-collector/blob/dc597d2316477c208f98903403f1d5f8c7ed5abf/Makefile.Common#L62) and does what I explained above. `make generate` is defined in the [Makefile](https://github.com/liatrio/liatrio-otel-collector/blob/dc597d2316477c208f98903403f1d5f8c7ed5abf/Makefile#L53) and basically runs `make gen` AKA `go generate ./...` on every location with a discovered [go.mod](https://github.com/liatrio/liatrio-otel-collector/blob/dc597d2316477c208f98903403f1d5f8c7ed5abf/Makefile#L8) file.

The basic idea is to run `make generate` after every code change, _always_. It'll either do nothing or create necessary changes. The documentation should reflect this

jburns24 replied:
Should the compgen Makefile include the root Makefile and not Makefile.common?

jburns24 replied:
`mdatagen` is removed but currently using the `make gen` target from Makefile.Common since that is what the generated Makefile imports

adrielp replied:
@acramsay  is correct, it would be `make gen`. It should include `Makefile.Common` not `Makefile`. I presume that this is instructions for the very first time you populate this, because once you have your component, you're going to be using the w/e rules exist at the root. Additionally, when you run `make checks` at the root, it's going to run through all of this, and the workflows will catch it too.

adrielp reviewed (changes requested):
Made some comments based out of our call yesterday.
